examples/marketplace-contract/src/lib.rs

```diff
@@ -0,0 +1,595 @@
+}
+
+// Distributes Selling Price, Royalty & Commission amounts.
+fn distribute_amounts<
+    S: HasStateApi,
+    T: IsTokenId + Clone + Copy,
+    A: IsTokenAmount + Clone + Copy + ops::Sub<Output = A> + Default,
+>(
+    host: &mut impl HasHost<State<S, T, A>, StateApiType = S>,
+    amount: Amount,
+    token_owner: &AccountAddress,
+    token_royalty_state: &TokenRoyaltyState,
+    marketplace_owner: &AccountAddress,
+) -> Result<(), MarketplaceError> {
+    let amounts = calculate_amounts(
+        &amount,
+        &host.state().commission,
+        token_royalty_state.royalty,
+    );
+
+    host.invoke_transfer(token_owner, amounts.to_seller)
+        .map_err(|_| MarketplaceError::InvokeTransferError)?;
+
+    if amounts
+        .to_marketplace
+        .cmp(&Amount::from_micro_ccd(0))
+        .is_gt()
+    {
+        host.invoke_transfer(marketplace_owner, amounts.to_marketplace)
+            .map_err(|_| MarketplaceError::InvokeTransferError)?;
+    }
+
+    if amounts
+        .to_primary_owner
+        .cmp(&Amount::from_micro_ccd(0))
+        .is_gt()
+    {
+        host.invoke_transfer(&token_royalty_state.primary_owner, amounts.to_primary_owner)
+            .map_err(|_| MarketplaceError::InvokeTransferError)?;
+    };
+
+    Ok(())
+}
+
+/// Callculates the amounts (Commission, Royalty & Selling Price) to be distributed
+fn calculate_amounts(
+    amount: &Amount,
+    commission: &Commission,
+    royalty_percentage_basis: u16,
+) -> DistributableAmounts {
+    let commission_amount = amount
+        .mul(commission.percentage_basis.into())
+        .quotient_remainder(MAX_BASIS_POINTS.into());
+
+    let royalty_amount = amount
+        .mul(royalty_percentage_basis.into())
+        .quotient_remainder(MAX_BASIS_POINTS.into());
+
+    DistributableAmounts {
+        to_seller: amount
+            .subtract_micro_ccd(commission_amount.0.micro_ccd())
+            .subtract_micro_ccd(royalty_amount.0.micro_ccd()),
+        to_marketplace: commission_amount.0,
+        to_primary_owner: royalty_amount.0,
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        add, calculate_amounts,
+        cis2_client::{
+            BALANCE_OF_ENTRYPOINT_NAME, OPERATOR_OF_ENTRYPOINT_NAME, SUPPORTS_ENTRYPOINT_NAME,
+        },
+        list,
+        params::AddParams,
+        state::{Commission, State, TokenInfo, TokenListItem, TokenPriceState, TokenRoyaltyState},
+        ContractState, ContractTokenAmount, ContractTokenId,
+    };
+    use concordium_cis2::*;
+
+    use concordium_std::{test_infrastructure::*, *};
+
+    const ACCOUNT_0: AccountAddress = AccountAddress([0u8; 32]);
+    const ADDRESS_0: Address = Address::Account(ACCOUNT_0);
+    const NFT_CONTRACT_ADDRESS: ContractAddress = ContractAddress {
+        index: 1,
+        subindex: 0,
+    };
+    const MARKET_CONTRACT_ADDRESS: ContractAddress = ContractAddress {
+        index: 2,
+        subindex: 0,
+    };
+
+    #[concordium_test]
+    fn should_add_token() {
```

Comment:
for users it is very useful if each test has a short summary at the top as to what it is meant to be testing. Otherwise it is hard to understand, and also maintain.